### PR TITLE
Updated precision_recall_curve.py

### LINKED ISSUE #1284: Ensure that all core and contrib metrics can work in DDP
## 🚀 Feature

Tracking issue. All core metrics are working in DDP, except special case #1283.
The idea is to check all core and contrib metrics if work in DDP and open new issues if we need to work on some metric.

### PR DESCRIPTION
**Pull request by sayantan1410**
Fixes #1284 

Updated the file precision_recall_curve.py file.
For testing I have ran the tests in the tests/ignite/contrib/metrics/test_precision_recall_curve.py and all of them passed but haven't separately tested in distributed environment.
It could contain some issues though.

### CODE REVIEW

File: ignite/contrib/metrics/precision_recall_curve.py [hunk 1]
```diff
@@ -70,6 +59,22 @@ def sigmoid_output_transform(output):
     """
 
     def __init__(self, output_transform: Callable = lambda x: x, check_compute_fn: bool = False) -> None:
+        try:
+            from sklearn.metrics import cohen_kappa_score  # noqa: F401
```
Comment (sdesrozis) [inline ignite/contrib/metrics/precision_recall_curve.py:63]:
```suggestion
            from sklearn.metrics import precision_recall_curve # noqa: F401
```

File: ignite/contrib/metrics/precision_recall_curve.py [hunk 2]
```diff
@@ -70,6 +59,22 @@ def sigmoid_output_transform(output):
...
+        except ImportError:
+            raise RuntimeError("This contrib module requires sklearn to be installed.")
+        self.precision_recall_curve_compute_fn = self.precision_recall_curve_compute()
```
Comment (sdesrozis) [inline ignite/contrib/metrics/precision_recall_curve.py:66]: I don’t understand this line, it seems useless.

File: ignite/contrib/metrics/precision_recall_curve.py [hunk 3]
```diff
@@ -70,6 +59,22 @@ def sigmoid_output_transform(output):
...
         super(PrecisionRecallCurve, self).__init__(
-            precision_recall_curve_compute_fn, output_transform=output_transform, check_compute_fn=check_compute_fn
+            self.precision_recall_curve_compute_fn, output_transform=output_transform, check_compute_fn=check_compute_fn
```
Comment (sdesrozis) [inline ignite/contrib/metrics/precision_recall_curve.py:68]:
```suggestion
            self.precision_recall_curve_compute, output_transform=output_transform, check_compute_fn=check_compute_fn
```

why not using directly the function ?

**Comment by sdesrozis** [comment]
@sayantan1410 Thanks ! I left a few comments.

File: ignite/contrib/metrics/precision_recall_curve.py (hunk 2 above)
Comment (sayantan1410) [reply ignite/contrib/metrics/precision_recall_curve.py:66]: Initially I also directly passed it to the super function but then I was taking reference for this from the file cohen_kappa.py where this was followed, so I thought may be there's some significance.

Comment (sdesrozis) [reply ignite/contrib/metrics/precision_recall_curve.py]: I see. Although you didn't do exactly the same. To do so, you should define a wrapper in your function as it's done in cohen_kappa.py.

Comment (sdesrozis) [reply ignite/contrib/metrics/precision_recall_curve.py]: I see. The function returns a wrapper, let's do it like that.

**Comment by sayantan1410** [comment]
@sdesrozis Corrected !! Please check.

**Comment by sdesrozis** [comment]
Anyway, the issue is solved. Actually, the tests should be refactored as cohen kappa ones. You can see the difference, it misses ddp tests. Could you try do improve the tests ?

**Comment by sayantan1410** [comment]
Yes, Sure I can do that, no problem !!

**Comment by sdesrozis** [comment]
> Yes, Sure I can do that, no problem !!

Thanks ! It should be very similar to tests/ignite/contrib/metrics/test_cohen_kappa.py !

File: ignite/contrib/metrics/precision_recall_curve.py [hunk 4]
```diff
@@ -1,21 +1,10 @@
-from typing import Any, Callable, Tuple
+from typing import Callable
 
 import torch
 
 from ignite.metrics import EpochMetric
 
 
-def precision_recall_curve_compute_fn(y_preds: torch.Tensor, y_targets: torch.Tensor) -> Tuple[Any, Any, Any]:
```
Comment (vfdev-5) [inline ignite/contrib/metrics/precision_recall_curve.py:8]: @sayantan1410 I think you misunderstood my point that we discussed offline. This PR wont solve DDP issue. Please revert the change. The problem is not here but in the way how works `EpochMetric.compute` method with unusual outputs. Check in details how works https://github.com/pytorch/ignite/blob/master/ignite/contrib/metrics/average_precision.py and see its tests, for example: https://github.com/pytorch/ignite/blob/493d86dc0eae7ad1fb4fd703cff2e3c7230dfbbf/tests/ignite/contrib/metrics/test_average_precision.py#L160-L185

Try to understand how it works and why we do https://github.com/pytorch/ignite/blob/493d86dc0eae7ad1fb4fd703cff2e3c7230dfbbf/tests/ignite/contrib/metrics/test_average_precision.py#L177-L178

Next you can get back to precision_recall_curve and see that `precision_recall_curve_compute_fn` does not return a scalar as does AP function: https://github.com/pytorch/ignite/blob/493d86dc0eae7ad1fb4fd703cff2e3c7230dfbbf/ignite/contrib/metrics/average_precision.py#L8

This is the root of the problem.

Comment (sayantan1410) [reply ignite/contrib/metrics/precision_recall_curve.py:8]: @vfdev-5 Okay so the problem is that precision_recall_curve returns 3 arrays but compute_fn for epoch metric can process an array !!

**Comment by sayantan1410** [comment]
@vfdev-5 I have updated the epoch_metric, so that it can broadcast tuples. Let me know what to correct.

File: ignite/contrib/metrics/precision_recall_curve.py [hunk 5]
```diff
@@ -21,7 +21,6 @@ class PrecisionRecallCurve(EpochMetric):
     by accumulating predictions and the ground-truth during an epoch and applying
     `sklearn.metrics.precision_recall_curve <https://scikit-learn.org/stable/modules/generated/
     sklearn.metrics.precision_recall_curve.html#sklearn.metrics.precision_recall_curve>`_ .
-
```
Comment (sdesrozis) [inline ignite/contrib/metrics/precision_recall_curve.py:24]: Please revert these changes in the docstring.

File: ignite/metrics/epoch_metric.py [hunk 6]
```diff
@@ -157,8 +157,14 @@ def compute(self) -> float:
             result = self.compute_fn(_prediction_tensor, _target_tensor)
 
         if ws > 1:
-            # broadcast result to all processes
-            result = cast(float, idist.broadcast(result, src=0))
+            if isinstance(result, tuple):
+                l = len(result)
+                for i in range(l):
+                    result[i] = idist.broadcast(result[i], src=0)
```
Comment (sdesrozis) [inline ignite/metrics/epoch_metric.py:163]: Why not adding the cast ?

**Comment by sdesrozis** [comment]
@sayantan1410 Thanks for the update. You should modify the tests to validate your improvement. See this file tests/ignite/metrics/test_epoch_metric.py

File: ignite/metrics/epoch_metric.py (hunk 6 above)
Comment (sayantan1410) [reply ignite/metrics/epoch_metric.py:163]: It is already a pytorch tensor, so I couldn't figure out what to cast to. Sorry for the delay, was caught up in a few works.

File: ignite/contrib/metrics/precision_recall_curve.py (hunk 5 above)
Comment (sayantan1410) [reply ignite/contrib/metrics/precision_recall_curve.py:24]: Yes, I will no problem.

**Comment by sayantan1410** [comment]
> @sayantan1410 Thanks for the update. You should modify the tests to validate your improvement. See this file tests/ignite/metrics/test_epoch_metric.py

Okay will check it out.

File: ignite/metrics/epoch_metric.py [hunk 7]
```diff
@@ -157,8 +157,14 @@ def compute(self) -> float:
             result = self.compute_fn(_prediction_tensor, _target_tensor)
 
         if ws > 1:
-            # broadcast result to all processes
-            result = cast(float, idist.broadcast(result, src=0))
+            if isinstance(result, tuple):
```
[the remark this answers is not in the document]
Comment (sayantan1410) [reply ignite/metrics/epoch_metric.py:160]: @vfdev-5 Okay, now I got it, I will have to re-write the compute function in the precision recall curve class separately. Right? Sorry for the delay, was caught up in a few work.

File: ignite/contrib/metrics/precision_recall_curve.py [hunk 8]
```diff
@@ -73,3 +75,32 @@ def __init__(self, output_transform: Callable = lambda x: x, check_compute_fn: b
         super(PrecisionRecallCurve, self).__init__(
             precision_recall_curve_compute_fn, output_transform=output_transform, check_compute_fn=check_compute_fn
         )
+
+    def compute(self) -> float:
```
Comment (sdesrozis) [inline ignite/contrib/metrics/precision_recall_curve.py:79]: 3 values are returned. You must fix the function signature.

Comment (sayantan1410) [reply ignite/contrib/metrics/precision_recall_curve.py:79]: sorry, overlooked it.

**Comment by sdesrozis** [comment]
I suggest to dive into tests/ignite/metrics/test_epoch_metric.py to get the understanding.

In order to improve the actual test, you can get some insights using another EpochMetric as regession/median_absolute_error.py.

File: ignite/contrib/metrics/precision_recall_curve.py [hunk 9]
```diff
@@ -73,3 +75,32 @@ def __init__(self, output_transform: Callable = lambda x: x, check_compute_fn: b
         super(PrecisionRecallCurve, self).__init__(
             precision_recall_curve_compute_fn, output_transform=output_transform, check_compute_fn=check_compute_fn
         )
+
+    def compute(self) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
```
Comment (sdesrozis) [inline ignite/contrib/metrics/precision_recall_curve.py:79]: I think returned value type is float.

File: ignite/metrics/epoch_metric.py [hunk 10]
```diff
@@ -136,7 +136,7 @@ def update(self, output: Tuple[torch.Tensor, torch.Tensor]) -> None:
             except Exception as e:
                 warnings.warn(f"Probably, there can be a problem with `compute_fn`:\n {e}.", EpochMetricWarning)
 
-    def compute(self) -> float:
+    def compute(self) -> Any:
```
Comment (sdesrozis) [inline ignite/metrics/epoch_metric.py:139]: I’m not sure about this. Returned value is float.

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py [hunk 11]
```diff
@@ -124,3 +125,36 @@ def test_check_compute_fn():
...
+            prc.update((y_pred, y))
+
+        # gather y_pred, y
+        y_pred = idist.all_gather(y_pred)
+        y = idist.all_gather(y)
+
+        np_y = y.cpu().numpy()
+        np_y_pred = y_pred.cpu().numpy()
+
+        res = prc.compute()
```
Comment (sdesrozis) [inline tests/ignite/contrib/metrics/test_precision_recall_curve.py:158]: res should be a tuple of float

Comment (sdesrozis) [inline tests/ignite/contrib/metrics/test_precision_recall_curve.py:158]: res would be a tuple of float

**Comment by sdesrozis** [comment]
@sayantan1410 Have a look to add distributed tests for TPU, GPU, etc.

https://github.com/pytorch/ignite/blob/3d9ba50bba42438a63183b3db0d4cc3a5e9b01ac/tests/ignite/contrib/metrics/regression/test_median_relative_absolute_error.py#L200

File: ignite/contrib/metrics/precision_recall_curve.py (hunk 9 above)
Comment (sayantan1410) [reply ignite/contrib/metrics/precision_recall_curve.py:79]: But I have casted to torch.Tensor in line 102, I have done that as the input array is also torch.Tensor.

File: ignite/metrics/epoch_metric.py (hunk 10 above)
Comment (sayantan1410) [reply ignite/metrics/epoch_metric.py:139]: As I am over riding the function and the return type of the over ridden function in precision_recall_curve class is Tuple which is not a sub class of float that is why I changed it to 'Any'. You can see this once as well https://github.com/pytorch/ignite/runs/5408929040?check_suite_focus=true#step:12:12

Apart from this if it is better to cast the return value of compute function in precision_recall_curve class to float I will do that.

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py (hunk 11 above)
Comment (sayantan1410) [reply tests/ignite/contrib/metrics/test_precision_recall_curve.py:158]: Okay, I will change the isinstance condition.

**Comment by sayantan1410** [comment]
@sdesrozis I have modified the tests, let me know if something else is needed to modify.

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py [hunk 12]
```diff
@@ -124,3 +127,95 @@ def test_check_compute_fn():
 
     em = PrecisionRecallCurve(check_compute_fn=False)
     em.update(output)
+
+
+def _test_distrib_binary_input(device):
```
Comment (sdesrozis) [inline tests/ignite/contrib/metrics/test_precision_recall_curve.py:132]: This method does not test anything. As `_test` method is never called, the executed code is empty.

Please have a look here :
https://github.com/pytorch/ignite/blob/3d9ba50bba42438a63183b3db0d4cc3a5e9b01ac/tests/ignite/contrib/metrics/regression/test_median_absolute_error.py#L152

The code would be very similar.

**Comment by sdesrozis** [comment]
@sayantan1410 I left a comment to address. Thanks in advance.

**Comment by sayantan1410** [comment]
@sdesrozis Fixed it, let me know if it is done correctly.

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py [hunk 13]
```diff
@@ -124,3 +127,161 @@ def test_check_compute_fn():
...
+        def update(engine, i):
+            return (
+                y_preds[i * size : (i + 1) * size],
+                y_true[i * size : (i + 1) * size],
+            )
+
+        engine = Engine(update)
+
+        prc = PrecisionRecallCurve(device=metric_device)
+        prc.attach(engine, "mare")
```
Comment (sdesrozis) [inline tests/ignite/contrib/metrics/test_precision_recall_curve.py:190]:
```suggestion
        prc.attach(engine, "prc")
```

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py [hunk 14]
```diff
@@ -124,3 +127,161 @@ def test_check_compute_fn():
...
+
+        data = list(range(n_iters))
+        engine.run(data=data, max_epochs=n_epochs)
+
+        assert "mare" in engine.state.metrics
```
Comment (sdesrozis) [inline tests/ignite/contrib/metrics/test_precision_recall_curve.py:195]:
```suggestion
        assert "prc" in engine.state.metrics
```

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py [hunk 15]
```diff
@@ -124,3 +127,161 @@ def test_check_compute_fn():
...
+
+        res = engine.state.metrics["mare"]
```
Comment (sdesrozis) [inline tests/ignite/contrib/metrics/test_precision_recall_curve.py:197]:
```suggestion
        res = engine.state.metrics["prc"]
```

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py [hunk 16]
```diff
@@ -124,3 +127,161 @@ def test_check_compute_fn():
...
+
+        np_y_true = y_true.cpu().numpy().ravel()
+        np_y_preds = y_preds.cpu().numpy().ravel()
+
+        e = np.abs(np_y_true - np_y_preds) / np.abs(np_y_true - np_y_true.mean())
```
Comment (sdesrozis) [inline tests/ignite/contrib/metrics/test_precision_recall_curve.py:202]: This can’t work. You copied / pasted the formula from mare metric. Please update this to compare to precision_recall_curve from sklearn as it’s done in others tests.

**Comment by sdesrozis** [comment]
Un

> @sdesrozis Fixed it, let me know if it is done correctly.

Thanks, I suggested a correction to be made to be correct.

**Comment by sayantan1410** [comment]
Really sorry, fixing it soon.

File: ignite/contrib/metrics/precision_recall_curve.py (hunk 9 above)
Comment (sdesrozis) [reply ignite/contrib/metrics/precision_recall_curve.py:79]:
```suggestion
    def compute(self) -> Tuple[float, float, float]:
```

File: ignite/contrib/metrics/precision_recall_curve.py [hunk 17]
```diff
@@ -73,3 +75,32 @@ def __init__(self, output_transform: Callable = lambda x: x, check_compute_fn: b
...
+        precision = torch.zeros(len(self._predictions))
+        recall = torch.zeros(len(self._predictions))
+        thresholds = torch.zeros(len(self._predictions) - 1)
+        if idist.get_rank() == 0:
+            # Run compute_fn on zero rank only
+            precision, recall, thresholds = self.compute_fn(_prediction_tensor, _target_tensor)
+
+        if ws > 1:
+            # broadcast result to all processes
+            precision = cast(torch.Tensor, idist.broadcast(precision, src=0))
```
Comment (sdesrozis) [inline ignite/contrib/metrics/precision_recall_curve.py:102]:
```suggestion
            precision = cast(float, idist.broadcast(precision, src=0))
```

File: ignite/contrib/metrics/precision_recall_curve.py [hunk 18]
```diff
@@ -73,3 +75,32 @@ def __init__(self, output_transform: Callable = lambda x: x, check_compute_fn: b
...
+            recall = cast(torch.Tensor, idist.broadcast(recall, src=0))
```
Comment (sdesrozis) [inline ignite/contrib/metrics/precision_recall_curve.py:103]:
```suggestion
            recall = cast(float, idist.broadcast(recall, src=0))
```

File: ignite/contrib/metrics/precision_recall_curve.py (hunk 9 above)
Comment (sdesrozis) [reply ignite/contrib/metrics/precision_recall_curve.py:79]: Ok i realised that the values are array type.

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py [hunk 19]
```diff
@@ -124,3 +127,161 @@ def test_check_compute_fn():
...
+        y_pred = idist.all_gather(y_pred)
+        y = idist.all_gather(y)
+
+        np_y = y.cpu().numpy()
+        np_y_pred = y_pred.cpu().numpy()
+
+        res = prc.compute()
+        assert isinstance(res, Tuple)
+        assert PrecisionRecallCurve(np_y, np_y_pred) == pytest.approx(res)
+        for _ in range(3):
```
Comment (sdesrozis) [inline tests/ignite/contrib/metrics/test_precision_recall_curve.py:163]: Wrong indentation.

The `_test()` function is still never called.

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py [hunk 20]
```diff
@@ -124,3 +127,161 @@ def test_check_compute_fn():
...
+
+        assert "prc" in engine.state.metrics
+
+        precision, recall, thresholds = engine.state.metrics["prc"]
+
+        np_y_true = y_true.cpu().numpy().ravel()
+        np_y_preds = y_preds.cpu().numpy().ravel()
+
+        sk_precision, sk_recall, sk_thresholds = precision_recall_curve(np_y_true, np_y_preds)
+        assert pytest.approx(precision) == sk_precision
```
Comment (sdesrozis) [inline tests/ignite/contrib/metrics/test_precision_recall_curve.py:203]: You could add an assert about stray shapes

Comment (sayantan1410) [reply tests/ignite/contrib/metrics/test_precision_recall_curve.py:203]: Okay will do that !!

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py (hunk 19 above)
Comment (sayantan1410) [reply tests/ignite/contrib/metrics/test_precision_recall_curve.py:163]: Sorry correcting it!!

**Comment by sayantan1410** [comment]
Hey, Can you tell why all the tests are not done when autopep8 makes a commit.

**Comment by sdesrozis** [comment]
I don't know. Did you apply all the tools (as blake, flake, etc.) before pushing ?

**Comment by sayantan1410** [comment]
It is asking me to add a blank line between line 162 and 163 in test_precision_recall_curve.py but that has been added by the autopep8 commit just after that, but I cannot figure out how to run the check for that commit. The problem raised as I haven't formatted the code with lint before pushing it.

**Comment by sdesrozis** [comment]
@sayantan1410 I think your tests are failing somewhere. The test `test_precision_recall_curve::test_distrib_gloo_cpu_or_gpu` is the first that fails, and since it's a distributed one, it causes all others distributed tests to fail.

Did you try to launch the parallel tests on your own machine ?

**Comment by sdesrozis** [comment]
@sayantan1410 Could you fix the tests ? Results are not equal to sklearn. Thanks !

**Comment by sayantan1410** [comment]
Yes, and the distributed tests have passed as well locally in CPU, but I cannot run the tests on GPU as I do not have GPU in my machine. 


> @sayantan1410 I think your tests are failing somewhere. The test `test_precision_recall_curve::test_distrib_gloo_cpu_or_gpu` is the first that fails, and since it's a distributed one, it causes all others distributed tests to fail.
> 
> Did you try to launch the parallel tests on your own machine ?

**Comment by sdesrozis** [comment]
> Yes, and the distributed tests have passed as well locally in CPU, but I cannot run the tests on GPU as I do not have GPU in my machine. 

You didn't run distributed tests because they didn't work. No GPU needed. In the tests folder, you must run the run_cpu_tests.py locally instead of pytest. Explore the script to catch how run distributed tests 😉

**Review by sdesrozis — APPROVED** [review]
@sayantan1410 You did it ! Thanks a lot for this work !

**Comment by vfdev-5** [comment]
@sayantan1410 @sdesrozis there is an issue with this PR, check this GPU CI job: https://app.circleci.com/pipelines/github/pytorch/ignite/2510/workflows/ced3dff0-05c4-4c92-b695-9c325b48af93/jobs/7720

```
____________________________ test_distrib_nccl_gpu _____________________________

distributed_context_single_node_nccl = {'local_rank': 0, 'rank': 0, 'world_size': 1}

    @pytest.mark.distributed
    @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
    @pytest.mark.skipif(torch.cuda.device_count() < 1, reason="Skip if no GPU")
    def test_distrib_nccl_gpu(distributed_context_single_node_nccl):
    
        device = idist.device()
>       _test_distrib_compute(device)

tests/ignite/contrib/metrics/test_precision_recall_curve.py:251: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/ignite/contrib/metrics/test_precision_recall_curve.py:191: in _test_distrib_compute
    _test(y_pred, y, batch_size, idist.device())
tests/ignite/contrib/metrics/test_precision_recall_curve.py:168: in _test
    res = prc.compute()
ignite/contrib/metrics/precision_recall_curve.py:103: in compute
    precision, recall, thresholds = self.compute_fn(_prediction_tensor, _target_tensor)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

y_preds = tensor([1, 1, 0, 1, 1, 0, 1, 1, 0, 0], device='cuda:0')
y_targets = tensor([0, 1, 0, 1, 1, 1, 0, 1, 0, 0], device='cuda:0')

    def precision_recall_curve_compute_fn(y_preds: torch.Tensor, y_targets: torch.Tensor) -> Tuple[Any, Any, Any]:
        try:
            from sklearn.metrics import precision_recall_curve
        except ImportError:
            raise RuntimeError("This contrib module requires sklearn to be installed.")
    
>       y_true = y_targets.numpy()
E       TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

ignite/contrib/metrics/precision_recall_curve.py:16: TypeError
_________________________ test_distrib_gloo_cpu_or_gpu _________________________

distributed_context_single_node_gloo = {'local_rank': 0, 'rank': 0, 'world_size': 1}

    @pytest.mark.distributed
    @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
    def test_distrib_gloo_cpu_or_gpu(distributed_context_single_node_gloo):
    
        device = idist.device()
>       _test_distrib_compute(device)

tests/ignite/contrib/metrics/test_precision_recall_curve.py:260: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/ignite/contrib/metrics/test_precision_recall_curve.py:191: in _test_distrib_compute
    _test(y_pred, y, batch_size, idist.device())
tests/ignite/contrib/metrics/test_precision_recall_curve.py:168: in _test
    res = prc.compute()
ignite/contrib/metrics/precision_recall_curve.py:103: in compute
    precision, recall, thresholds = self.compute_fn(_prediction_tensor, _target_tensor)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

y_preds = tensor([1, 1, 0, 1, 1, 0, 1, 1, 0, 0], device='cuda:0')
y_targets = tensor([0, 1, 0, 1, 1, 1, 0, 1, 0, 0], device='cuda:0')

    def precision_recall_curve_compute_fn(y_preds: torch.Tensor, y_targets: torch.Tensor) -> Tuple[Any, Any, Any]:
        try:
            from sklearn.metrics import precision_recall_curve
        except ImportError:
            raise RuntimeError("This contrib module requires sklearn to be installed.")
    
>       y_true = y_targets.numpy()
E       TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

```

**Comment by sdesrozis** [comment]
I'm on it.

File: ignite/contrib/metrics/precision_recall_curve.py [hunk 21]
```diff
@@ -69,7 +71,48 @@ def sigmoid_output_transform(output):
...
-            precision_recall_curve_compute_fn, output_transform=output_transform, check_compute_fn=check_compute_fn
+            precision_recall_curve_compute_fn,
+            output_transform=output_transform,
+            check_compute_fn=check_compute_fn,
+            device=device,
         )
+
+    def compute(self) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        if len(self._predictions) < 1 or len(self._targets) < 1:
+            raise NotComputableError("EpochMetric must have at least one example before it can be computed.")
```
Comment (vfdev-5) [inline ignite/contrib/metrics/precision_recall_curve.py:89]:
```diff
- "EpochMetric must have ..."
+ "PrecisionRecallCurve must have ..."
```

File: ignite/contrib/metrics/precision_recall_curve.py [hunk 22]
```diff
@@ -69,7 +71,48 @@ def sigmoid_output_transform(output):
...
+        self._is_reduced = True
+
+        if idist.get_rank() == 0:
+            # Run compute_fn on zero rank only
+            precision, recall, thresholds = self.compute_fn(_prediction_tensor, _target_tensor)
+            precision = torch.Tensor(precision)
+            recall = torch.Tensor(recall)
+            # thresholds can have negative strides, not compatible with torch tensors
+            # https://discuss.pytorch.org/t/negative-strides-in-tensor-error/134287/2
+            thresholds = torch.Tensor(thresholds.copy())
```
Comment (vfdev-5) [inline ignite/contrib/metrics/precision_recall_curve.py:108]: Tensor creation should be done with `torch.tensor` and not `torch.Tensor`

**Review by vfdev-5 — COMMENTED** [review]
@sayantan1410 thanks for the PR and sorry for delay. Please send another one to fix my review comments.

File: ignite/contrib/metrics/precision_recall_curve.py (hunk 22 above)
Comment (sayantan1410) [reply ignite/contrib/metrics/precision_recall_curve.py:108]: Okay, will change it.

File: ignite/contrib/metrics/precision_recall_curve.py (hunk 21 above)
Comment (sayantan1410) [reply ignite/contrib/metrics/precision_recall_curve.py:89]: Sorry, Fixing it soon.

> @sayantan1410 You did it ! Thanks a lot for this work !

Thank you for the help, got to learn a lot.

Comment (sayantan1410) [reply ignite/contrib/metrics/precision_recall_curve.py:89]: Fixing it soon.

**Comment by sayantan1410** [comment]
@sdesrozis Thank you for all the help, got to learn a lot !!

File: tests/ignite/contrib/metrics/test_precision_recall_curve.py [hunk 23]
```diff
@@ -124,3 +135,182 @@ def test_check_compute_fn():
...
+        test_cases = get_test_cases()
+        for y_pred, y, batch_size in test_cases:
+            _test(y_pred, y, batch_size, "cpu")
+            if device.type != "xla":
+                _test(y_pred, y, batch_size, idist.device())
+
+
+def _test_distrib_integration(device):
+
+    rank = idist.get_rank()
```
Comment (vfdev-5) [inline tests/ignite/contrib/metrics/test_precision_recall_curve.py:196]: This test case does not use rank and each process generates some random preds and true values. Implementation of `PrecisionRecallCurve` should gather all data from all processes but it is checked against local process computation:
```
        np_y_true = y_true.cpu().numpy().ravel()
        np_y_preds = y_preds.cpu().numpy().ravel()
        sk_precision, sk_recall, sk_thresholds = precision_recall_curve(np_y_true, np_y_preds)
```

I would assume this to fail but looks like it is passing. @sayantan1410 can you check why it is so ?

Comment (sayantan1410) [reply tests/ignite/contrib/metrics/test_precision_recall_curve.py:196]: Yeah sur will check it !!

Comment (sdesrozis) [reply tests/ignite/contrib/metrics/test_precision_recall_curve.py:196]: Actually the error comes from the following which seems wrong 
https://github.com/pytorch/ignite/blob/90d79aa7b6d0a52b2741943b62baab2ea0461453/tests/ignite/contrib/metrics/regression/test_median_absolute_error.py#L152

@sayantan1410 Have a look to this correct implementation
https://github.com/pytorch/ignite/blob/90d79aa7b6d0a52b2741943b62baab2ea0461453/tests/ignite/contrib/metrics/regression/test_mean_error.py#L139

Comment (sayantan1410) [reply tests/ignite/contrib/metrics/test_precision_recall_curve.py:196]: @sdesrozis A small question, 
https://github.com/pytorch/ignite/blob/90d79aa7b6d0a52b2741943b62baab2ea0461453/tests/ignite/contrib/metrics/regression/test_mean_error.py#L139

Here, also we don't have something like idist.all_gather so how is the data getting gathered from all the processes ?
